### PR TITLE
Par process entries

### PR DESCRIPTION
**Pull request by aeyakovenko**
Execute entries in parallel in the replicate stage.

### CODE REVIEW

File: src/bank.rs [hunk 1]
```diff
@@ -912,9 +913,60 @@ impl Bank {
...
+            }).collect();
+        Self::all_ok(&results)
+    }
+
+    /// process entries in parallel
+    /// 1. In order lock accounts for each entry while the lock succeeds, up to a Tick entry
+    /// 2. Process the locked group in parallel
+    /// 3. Register the `Tick` if it's available, goto 1
+    pub fn par_process_entries(&self, entries: &[Entry]) -> Result<()> {
+        // accumilator for entries that can be processed in parallel
```
Comment (sakridge) [inline src/bank.rs]: accumulator

File: src/bank.rs [hunk 2]
```diff
@@ -912,9 +913,60 @@ impl Bank {
 
     /// Process an ordered list of entries.
     pub fn process_entries(&self, entries: &[Entry]) -> Result<()> {
+        self.par_process_entries(entries)
+    }
+
+    pub fn all_ok(results: &[Result<()>]) -> Result<()> {
+        for r in results {
```
Comment (garious) [inline src/bank.rs]: Seems like someone would have created a library function for this, and it'd look something like this:

```rust
pub fn first_err<T: Default, E>(results: Vec<Result<T, E>>) -> Result<T, E> {
     results.into_iter().fold(Ok(Default::default()), Result.and)
}
```

Comment (garious) [reply src/bank.rs]:
```rust
#[derive(Debug, PartialEq)]
enum E {
    E0,
    E1,
}

#[test]
fn test_first_err() {
    assert_eq!(first_err::<(), E>(vec![]), Ok(()));
    assert_eq!(first_err::<(), E>(vec![Ok(())]), Ok(()));
    assert_eq!(first_err::<(), _>(vec![Err(E::E0)]), Err(E::E0));
    assert_eq!(first_err::<(), _>(vec![Err(E::E0), Err(E::E1)]), Err(E::E0));
}
```

Comment (sakridge) [reply src/bank.rs]: +1 for the first_err name, I was going to comment as well all_ok doesn't indicate what it does that well.

Comment (garious) [reply src/bank.rs]: Sunday fun day

```rust
trait ResultIterator<T, E> {
    fn first_err(self) -> Result<T, E>;
}

impl<I, T, E> ResultIterator<T, E> for I
where
    I: IntoIterator<Item = Result<T, E>>,
    T: Default,
{
    fn first_err(self) -> Result<T, E> {
        self.into_iter().fold(Ok(Default::default()), Result::and)
    }
}

#[derive(Debug, PartialEq)]
enum E {
    E0,
    E1,
}

#[test]
fn test_first_err() {
    let empty: Vec<Result<_, E>> = vec![];
    assert_eq!(empty.first_err(), Ok(()));

    let one_ok: Vec<Result<_, E>> = vec![Ok(())];
    assert_eq!(one_ok.first_err(), Ok(()));

    let one_err: Vec<Result<(), _>> = vec![Err(E::E0)];
    assert_eq!(one_err.first_err(), Err(E::E0));

    let two_errs: Vec<Result<(), _>> = vec![Err(E::E0), Err(E::E1)];
    assert_eq!(two_errs.first_err(), Err(E::E0));

    assert_eq!(vec![Ok(()), Err(E::E0)].first_err(), Err(E::E0));
}
```

File: src/bank.rs [hunk 3]
```diff
@@ -912,9 +913,60 @@ impl Bank {
...
+            if r.is_err() {
```
Comment (garious) [inline src/bank.rs]: Alternatively, change this to `r?;` :)

File: src/bank.rs (hunk 2 above)
Comment (garious) [reply src/bank.rs]: So we never need to think about this again, can you add this to result.rs:

```rust
pub fn first_err<I, T, E>(xs: I, def: T) -> Result<T, E>
where
    I: IntoIterator<Item = Result<T, E>>,
{
    for x in xs {
        x?;
    }
    Ok(def)
}

#[derive(Debug, PartialEq)]
enum E {
    E0,
    E1,
}

#[test]
fn test_first_err() {
    assert_eq!(first_err::<_, _, E>(vec![], ()), Ok(()));
    assert_eq!(first_err::<_, _, E>(vec![Ok(())], ()), Ok(()));
    assert_eq!(first_err(vec![Err(E::E0)], ()), Err(E::E0));
    assert_eq!(first_err(vec![Err(E::E0), Err(E::E1)], ()), Err(E::E0));
    assert_eq!(first_err(vec![Ok(()), Err(E::E0)], ()), Err(E::E0));
}
```

Comment (aeyakovenko) [reply src/bank.rs]: @garious want me to use this?  or you think its overkill?

Comment (garious) [reply src/bank.rs]: Use it.  It's not overkill. Add it to result.rs.  I've used this `for x in xs { x? }` pattern a dozen times, and I'd like to finally have a function for it.

**Comment by garious** [comment]
I vote no on this PR. No tests, no benchmarks, and no explanation for why the system would benefit from validators that outperform the leader despite the increase in code complexity.

**Comment by aeyakovenko** [comment]
@garious the validators are not keeping up with 200ktps.  same process_entries tests cover this change.

**Comment by garious** [comment]
@aeyakovenko, is it only the case that they're not keeping up when running on lesser hardware? Once leader rotation is in place, will we want to support that scenario?  Anyway, offering #1502 as a possible, simpler, alternative.

**Comment by garious** [comment]
@aeyakovenko, can you review the coverage report on Buildkite and add tests for your new functions and their branches? https://solana-ci-gate.herokuapp.com/buildkite_public_artifact?https://api.buildkite.com/v2/organizations/solana-labs/pipelines/solana/builds/4052/jobs/43400d92-3cb7-4ca9-9041-72e156940eef/artifacts/1040b0e0-a00d-4da6-949e-e4ee2b1eb8fd/download.

**Comment by aeyakovenko** [comment]
@garious awesome!  how do we get that data into PRs?

File: src/bank.rs [hunk 4]
```diff
@@ -1886,4 +1940,90 @@ mod tests {
...
+            Bank::first_err(&[
+                Err(BankError::AccountInUse),
+                Ok(()),
+                Err(BankError::DuplicateSignature)
+            ]),
+            Err(BankError::AccountInUse)
+        );
+    }
+    #[test]
+    fn test_par_process_entries() {
```
Comment (garious) [inline src/bank.rs]: Can you split this into 4 tests? I don't mind the duplication of those first 4 lines. It's easier to grok than figuring out where one test depends on data or side-effects of the previous.

File: src/bank.rs [hunk 5]
```diff
@@ -912,9 +913,62 @@ impl Bank {
...
+    }
+    pub fn par_execute_entries(&self, entries: &[(&Entry, Vec<Result<()>>)]) -> Result<()> {
+        inc_new_counter_info!("bank-par_execute_entries-count", entries.len());
+        let results: Vec<Result<()>> = entries
+            .into_par_iter()
+            .map(|(e, locks)| {
+                let results = self.execute_and_commit_transactions(
+                    &e.transactions,
+                    locks.to_vec(),
+                    MAX_ENTRY_IDS,
```
Comment (ryoqun) [inline src/bank.rs:933]: For the record, this should be [`MAX_ENTRY_IDS / 2` like the other callsite](https://github.com/solana-labs/solana/pull/1499/files#diff-7145777648f8c12caeb15b4e1755215cL718), we fixed this almost 1.5 years later #9423. ;)

**Pull request merged**